src/paginator-spec.js

```diff
@@ -0,0 +1,486 @@
+      });
+    });
+
+    describe('constructor', function() {
+      const givenPageNumber = 42;
+
+      beforeEach(function() {
+        const queryParams = {
+          page: givenPageNumber
+        };
+        paginator = new Paginator(request.spy, null, queryParams);
+      });
+
+      it('sets the page to the given page query parameter', function() {
+        expect(paginator.page).to.equal(givenPageNumber);
+      });
+    });
+
+    describe('constructor', function() {
+      const message = Paginator.errors.invalidStyle;
+      const style = 'invalid-value';
+
+      it('throws an error when an invalid pagination style is set',
+        function() {
+          const paginatorConstructor = function() {
+            new Paginator(request.spy, null, {}, { style });
+          };
+
+          return expect(paginatorConstructor).to.throw(Error, message);
+        });
+    });
+
+    describe('constructor', function() {
+      describe('(LimitOffsetPagination)', function() {
+        const style = Paginator.styles.limitOffset;
+
+        it('throws an error when no limit is set and offset isn\'t zero',
+          function() {
+            const message = Paginator.errors.calculatePageNoLimit;
+            const queryParams = { offset: 5 };
+            const paginatorConstructor = function() {
+              new Paginator(request.spy, null, queryParams, { style });
+            };
+
+            return expect(paginatorConstructor).to.throw(Error, message);
+          });
+
+        it('doesn\'t throw an error when no limit is set and offset is zero',
+          function() {
+            const message = Paginator.errors.calculatePageNoLimit;
+            const queryParams = { offset: 0 };
+            const paginatorConstructor = function() {
+              new Paginator(request.spy, null, queryParams, { style });
+            };
+
+            return expect(paginatorConstructor).to.not.throw(Error, message);
+          });
+      });
+    });
+
+    describe('page', function() {
+      beforeEach(function() {
+        paginator = new Paginator(request.spy);
+      });
+
+      it('defaults to one', function() {
+        expect(paginator.page).to.equal(1);
+      });
+    });
+
+    describe('next', function() {
+      beforeEach(function() {
+        paginator = new Paginator(request.spy);
+      });
+
+      it('returns a promise that successfully resolves', function() {
+        const promise = paginator.next();
+
+        return expect(promise).to.eventually.resolve;
+      });
+
+      it('returns a promise with a response instance', function() {
+        const response = paginator.next();
+
+        return expect(response).to.eventually.be.an.instanceOf(Response);
+      });
+
+      it('increments the current page', function() {
+        expect(paginator.page).to.equal(1);
+
+        const currentPage = paginator.next().then(getCurrentPage);
+
+        return expect(currentPage).to.eventually.equal(2);
```

Comment:
This is wrong. The first call to `next` should return a promise to the current page. Subsequent calls should increment it.

We are purposefully implementing this similar to an iterator, in which the first call to `next` returns the first value. `first` should be considered a convenience method and not the primary way to get the first page. 

The first element shouldn't require any special treatment. Consider:

```js
let firstFivePagePromises = [];

firstFivePagePromises.push(paginator.first());

for (let i= 0; i < 4; i++) {
  firstFivePagePromises.push(paginator.next());
}
```

vs

```js
let firstFivePagePromises = [];

for (let i= 0; i < 5; i++) {
  firstFivePagePromises.push(paginator.next());
}
```

For `first`, it it's to reset the paginator to the beginning, it should set it to page `0`;